Voigt kampff: Bugfix dialogfile from sentence

forslund opened the pull request:
## Description
Fix two issues with `"skill" should reply with "sentence"` behave step.

- locale folder was not checked for dialog files
- A redundant (?) regex step caused issues if there were no keywords

Should resolve #3067

## How to test
Ensure VK tests still works, see test case in #3067 for previously failing test.

## Contributor license agreement signed?
CLA [ x ]

forslund committed:
VK: Include locale folders in dialog_from_sentence

Only dialog folder was used previously this adds globbing through the
locale/lang/ folder and it's subfolders

forslund committed:
VK: Fix regex used in _match_dialog_patterns

A redundant step caused issue when performing tests in the tv-remove-control-skill

devvmh commented on test/integrationtests/voight_kampff/features/steps/utterance_responses.py:
```diff
@@ -65,6 +65,26 @@ def load_dialog_list(skill_path, dialog):
+        skill_path (str): skill root folder
+        lang (str): language code to check
+
+    yeilds:
```
nit: typo

forslund replied:
Thanks will fix

devvmh commented on test/integrationtests/voight_kampff/features/steps/utterance_responses.py:
```diff
@@ -65,6 +65,26 @@ def load_dialog_list(skill_path, dialog):
+    yeilds:
+        (Path) path of each found dialog file
+    """
+    in_dialog_dir = Path(skill_path, 'dialog', lang).rglob('*.dialog')
+    for dialog_path in in_dialog_dir:
+        yield dialog_path
```
should it search locale first or dialog first?

forslund replied:
It doesn't really matter as both should generally not exist in the same skill so I went alphabetically :)

forslund replied:
If you think it's more logical to do it in the reverse order I can update that.

devvmh replied:
yeah, you're right, either seems fine

devvmh commented on test/integrationtests/voight_kampff/features/steps/utterance_responses.py:
```diff
@@ -75,9 +95,9 @@ def dialog_from_sentence(sentence, skill_path, lang):
     best = (None, 0)
-    for path in glob(dialog_paths):
+    for path in _get_dialog_files(skill_path, lang):
+        print('Path', path)
```
oh whoops can revert this

forslund replied:
Doh, thought I removed that. will fix

devvmh commented:
i've applied https://github.com/forslund/mycroft-core/commit/03ef7f4b219533507f016c64a45487bc827ec78c and https://github.com/forslund/mycroft-core/commit/825b3879a51c15513e069e61872266bfa55ecd99 locally but don't have it working yet - i must need to refresh a cache or something.

assuming it works, i would love to see this backported to master since i'm not sure i'm ready to move to dev branch... but i can struggle along with cherry-picking it otherwise

devvmh commented on test/integrationtests/voight_kampff/features/steps/utterance_responses.py:
```diff
@@ -98,7 +117,6 @@ def _match_dialog_patterns(dialogs, sentence):
     dialogs = [re.sub(r'{.*?\}', r'.*', dia) for dia in dialogs]
     # Remove left over '}'
     dialogs = [re.sub(r'\}', r'', dia) for dia in dialogs]
-    dialogs = [re.sub(r' .* ', r' .*', dia) for dia in dialogs]
```
ok i think this is actually a bug. 

my dialog file:

```
(OK | ) turning on the TV
```

My test file:

```
Feature: turn-on-tv
  Scenario: Turning on the TV
    Given an English speaking user
    When the user says "turn on the tv"
    Then "tv-remote-control-skill" should reply with "turning on the tv"
```

When I test with this, I get

```
Dialogs4: ['^(ok | ) turning on the tv']
```

after this line. I think it should probably be something like `'.* turning on the tv'`, right?

I thought I had this working locally but don't anymore. Will keep trying

devvmh replied:
or maybe this never worked?

devvmh replied:
oh I put it in the wrong order. this is what works

code:

```
    # Allow custom fields to be anything
    dialogs = [re.sub(r'{.*?\}', r'.*', dia) for dia in dialogs]
    print("Dialogs1: {}".format(dialogs))
    # Remove left over '}'
    dialogs = [re.sub(r'\}', r'', dia) for dia in dialogs]
    print("Dialogs2: {}".format(dialogs))
    dialogs = [re.sub(r' [^\)]* ', r' .*', dia) for dia in dialogs]
    print("Dialogs3: {}".format(dialogs))
    # Merge consequtive .*'s into a single .*
    dialogs = [re.sub(r'\.\*( \.\*)+', r'.*', dia) for dia in dialogs]
    print("Dialogs4: {}".format(dialogs))
    # Remove double whitespaces
    dialogs = ['^' + ' '.join(dia.split()) for dia in dialogs]
    print("Dialogs5: {}".format(dialogs))
```

output:

```
      Dialogs1: ['(ok | ) turning on the tv']
      Dialogs2: ['(ok | ) turning on the tv']
      Dialogs3: ['(ok .*) .*tv']
      Dialogs4: ['(ok .*) .*tv']
      Dialogs5: ['^(ok .*) .*tv']
```

devvmh replied:
I'd argue though at this point the regex might be harming more than it's helping

1. Hard to understand/maintain
2. Seems to have bugs e.g. it's stripping `turning on the` from my dialog. 

Might be better to expand this to a set of phrases and then just match with a simpler regex. Or is there a way to leverage the code that actually parses these files? If you could share a module between this code and that code it would guarantee that updates to the dialog syntax are always reflected in the test behaviour.

forslund replied:
I didn't realize the parenthesis expansion was run on the dialogfiles (or maybe I've forgotten), and yes it should be expanded using the normal parenthesis expansion. The regex intention is to handle the data entries:
`hello dear {person_type}`. and replace the `{person_type}` with `.*` since that could be anything. There doesn't really exist any dialog matching system anywhere since it's not an input to Mycroft.

Will try to improve it.

forslund replied:
Hmm, and to further improve this I think we can do like you say and use the normal dialog renderer to render the final regex in one step...(or maybe two steps)

forslund replied:
I did a small change to expand parentheses in the dialog lines, which made your example work for me(TM). I will attempt a bigger change tonight to try to remove the big weird multi-pass regex (and fix the pep8speaks issue)

forslund replied:
I did another pass at it using the normal dialog loader to remove some of the regex clutter together with using the built in string format-parser...

Need to test it some more tomorrow morning before work.

devvmh replied:
Love the look of the new approach. It seems to be working now but check my other comment

forslund committed:
VK: Expand the parentheses from dialog files

pep8speaks commented:
Hello @forslund! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2022-02-23 06:01:56 UTC

forslund committed:
VK simplify dialog matching

This uses the existing dialog renderer and the standard format library
to in two steps create a regex where the {elements} in a dialog is
replaced with ".*" to match the given sentence.

devvmh commented on test/integrationtests/voight_kampff/features/steps/utterance_responses.py:
```diff
@@ -92,17 +119,23 @@ def dialog_from_sentence(sentence, skill_path, lang):
+
+    # Remove double whitespaces and ensure that it matches from
+    # the beginning of the line.
+    regexes = ['^' + ' '.join(reg.split()) for reg in regexes]
```
unused? I'm trying to debug through why this works but doesn't output anything... strange.

forslund replied:
I think the start of line anchor is usable but the split may not be. (I thought the expand_options could leave extra ' ' but it seems it doesn't). I haven't have time to do many more tests today so it'll be a weekend project.

I'll remove the space cleanup, thanks for reviewing and finding improvements!

devvmh replied:
Right I just meant the regexes variable isn't consumed later in the function.

Thanks for working on this issue!

forslund replied:
*face palm* doh, right you are. That would have screwed up the pattern matching of data entries completely. thank you!

I've amended the offending commit. but will likely do the actual testing tomorrow.

devvmh replied:
OK! I pulled the amended commit and at least for my limited set of dialog files, it's working correctly.

forslund replied:
The tests I did also looks good.

devvmh reviewed (approved):
Not sure if my review counts but I say ship it

krisgesling commented:
Thanks @devvmh your review is very helpful!

I've been away having a baby and covid... but hopefully all our CI tests are happy now and we can merge it in :tada:

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-3068))

krisgesling reviewed (approved):
:ship: :rocket:
